src/main/java/core/basesyntax/Car.java

```diff
@@ -1,13 +1,121 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
-/**
- * Make this class immutable. See requirements in the README file
- */
-public class Car {
-    private int year;
-    private String color;
-    private List<Wheel> wheels;
-    private Engine engine;
+public final class Car implements Cloneable {
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        if (wheels == null) {
+            throw new NullPointerException();
+        }
+
+        if (engine == null) {
+            this.engine = null;
+        } else {
+            this.engine = engine.clone();
+        }
+
+        this.year = year;
+        this.color = color;
+        this.wheels = cloneWheels(wheels);
+    }
+
+    public int getYear() {
+        return year;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public List<Wheel> getWheels() {
+        return cloneWheels(wheels);
+    }
+
+    public Engine getEngine() {
+        if (engine == null) {
+            return null;
+        }
+
+        return engine.clone();
+    }
+
+    public Car changeEngine(Engine engine) {
+        if (engine == null) {
+            return null;
+        }
+
+        return new Car(year, color, wheels, engine);
+    }
+
+    public Car addWheel(Wheel wheel) {
+        if (wheel == null) {
+            return null;
+        }
+
+        List<Wheel> wheels = cloneWheels(getWheels());
```

Comment:
```suggestion
        List<Wheel> wheels = cloneWheels(wheels);
```
no need to copy wheels twice in get method and cloneWheels